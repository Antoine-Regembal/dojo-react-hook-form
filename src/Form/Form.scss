.form {
  padding: 3rem 0 0 3rem;

  &__input {
    padding: 0.3rem 0.8rem 0.3rem 0.8rem;
  }
  &__button {
    margin-top: 3rem;
    padding: 0.3rem 0.8rem 0.3rem 0.8rem;
  }
}

.input-wrapper {
  display: flex;
  justify-content: space-between;
  width: 60%;
  margin: .5rem;
}
.input-wrapper:not(:first-of-type) {
  margin-top: 2rem;
}