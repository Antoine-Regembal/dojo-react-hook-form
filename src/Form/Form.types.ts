export interface FieldValueUpdate {
  action: "valueUpdate" | "errorsUpdate";
  field: string;
  value: string | string[];
}

export interface IFieldStates {
  value: string;
  errors: string[];
}

export interface IInitialFormValues {
  firstname: IFieldStates;
  lastname: IFieldStates;
  email: IFieldStates;
  confirmEmail: IFieldStates;
}

export interface IFieldValidationRules {
  [ruleKey: string]: string | boolean;
}

export type DispatchFormValue<T, Y> = (state: T, fieldUpdate: Y) => T;
