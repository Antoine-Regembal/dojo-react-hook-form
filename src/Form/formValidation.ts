import React from "react";
import {
	IFormErrors,
	IInitialFormValues,
	IFieldValidationRules
} from "./Form.types";

export const fieldsValidationRules = {
	firstname: {
		required: true
	},
	lastname: {
		required: true
	},
	email: {
		required: true,
		regexValidation: "^[\\w-.]+@([\\w-]+\\.)+[\\w-]{2,4}$"
	},
	confirmEmail: {
		required: true,
		equalsTo: "email"
	}
};

const addFieldError = (
	setFormErrors: React.Dispatch<React.SetStateAction<IFormErrors>>,
	field: string,
	errorMessage: string
) => {
	setFormErrors((prevFieldsErrorsState: IFormErrors) => {
		return {
			...prevFieldsErrorsState,
			[field]: [
				...new Set([...prevFieldsErrorsState[field as keyof IFormErrors], errorMessage]),
			],
		};
	});
};

const removeFieldError = (
	setFormErrors: React.Dispatch<React.SetStateAction<IFormErrors>>,
	field: string,
	errorMessage: string
) => {
	setFormErrors((prevFieldsErrorsState: IFormErrors) => {
		return {
			...prevFieldsErrorsState,
			[field]: [
				...new Set([
					...prevFieldsErrorsState[field as keyof IFormErrors].filter(
						(currErrorMessage: string) => currErrorMessage !== errorMessage
					),
				]),
			],
		};
	});
};

const isFieldValueEmpty = (fieldValue: string) => fieldValue === undefined || fieldValue === null || fieldValue === "";

export const validateFieldRules = (
	field: string,
	value: string,
	fieldValidationRules: IFieldValidationRules,
	setFormErrors: React.Dispatch<React.SetStateAction<IFormErrors>>,
	formValues?: IInitialFormValues
) => {
	Object.entries(fieldValidationRules).forEach(([ruleName, ruleValue]) => {
		switch (ruleName) {
		case "required":
			if (ruleValue) {
				isFieldValueEmpty(value)
					? addFieldError(setFormErrors, field, "Field cannot be empty")
					: removeFieldError(setFormErrors, field, "Field cannot be empty");
			}
			break;
		case "regexValidation":
			!value.match(ruleValue as string)
				? addFieldError(setFormErrors, field, "Invalid format")
				: removeFieldError(setFormErrors, field, "Invalid format");
			break;
		case "equalsTo":
			!(value === formValues?.[ruleValue as keyof IInitialFormValues])
				? addFieldError(setFormErrors, field, "Does not match")
				: removeFieldError(setFormErrors, field, "Does not match");
			break;
		default:
			break;
		}
	});
};

export const validateForm = (
	formValues: IInitialFormValues,
	formErrors: IFormErrors,
	setFormErrors: React.Dispatch<React.SetStateAction<IFormErrors>>
) => {
	Object.entries(formValues).forEach(([field, value]: [field: string, value: string]) => {
		const fieldValidationRules = fieldsValidationRules[field as keyof typeof fieldsValidationRules];

		validateFieldRules(
			field,
			value,
			fieldValidationRules,
			setFormErrors,
			formValues
		);
	});
};
