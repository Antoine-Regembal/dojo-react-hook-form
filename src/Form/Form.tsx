import React, {useReducer, useState} from "react";
import "./Form.scss";

import {DispatchFormValue, FieldValueUpdate, IInitialFormValues, IFormErrors} from "./Form.types";

import {validateFieldRules, validateForm, fieldsValidationRules} from "./formValidation";

export const Form = () => {
	const dispatchFormValue = (state: IInitialFormValues, fieldValueUpdate: FieldValueUpdate) => {
		return {...structuredClone(state), [fieldValueUpdate.field]: fieldValueUpdate.value} as IInitialFormValues;
	};
	const [formValues, setFormValues] = useReducer<DispatchFormValue<IInitialFormValues, FieldValueUpdate>>(
		dispatchFormValue,
		{
			firstname: "",
			lastname: "",
			email: "",
			confirmEmail: ""
		}
	);

	const [formErrors, setFormErrors] = useState<IFormErrors>({
		firstname: [],
		lastname: [],
		email: [],
		confirmEmail: []
	});

	const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
		e.preventDefault();

		validateForm(formValues, formErrors, setFormErrors);
	};

	return (
		<form className="form form--required" onSubmit={e => handleSubmit(e)}>
			<div className="input-wrapper">
				<label className="form__label form__label--required" htmlFor="firstname">Firstname</label>
				<input
					className={`form__input${formErrors.firstname.length ? " form__input--error" : ""}`}
					id="firstname"
					placeholder="Firstname"
					type="text"
					value={formValues.firstname}
					onBlur={() => validateFieldRules(
						"firstname",
						formValues.firstname,
						fieldsValidationRules.firstname,
						setFormErrors
					)}
					onChange={e => setFormValues({field: "firstname", value: e.target.value})}/>
				<em role="alert" className="form__error">{formErrors.firstname.join(" - ")}</em>
			</div>
			<div className="input-wrapper">
				<label className="form__label form__label--required" htmlFor="lastname">Lastname</label>
				<input
					className={`form__input${formErrors.lastname.length ? " form__input--error" : ""}`}
					id="lastname"
					placeholder="Lastname"
					type="text"
					value={formValues.lastname}
					onBlur={() => validateFieldRules(
						"lastname",
						formValues.lastname,
						fieldsValidationRules.lastname,
						setFormErrors
					)}
					onChange={e => setFormValues({field: "lastname", value: e.target.value})}/>
				<em role="alert" className="form__error">{formErrors.lastname.join(" - ")}</em>
			</div>
			<div className="input-wrapper">
				<label className="form__label form__label--required" htmlFor="email">Email</label>
				<input
					className={`form__input${formErrors.email.length ? " form__input--error" : ""}`}
					id="email"
					placeholder="Email"
					type="text"
					value={formValues.email}
					onBlur={() => validateFieldRules(
						"email",
						formValues.email,
						fieldsValidationRules.email,
						setFormErrors
					)}
					onChange={e => setFormValues({field: "email", value: e.target.value})}/>
				<em role="alert" className="form__error">{formErrors.email.join(" - ")}</em>
			</div>
			<div className="input-wrapper">
				<label className="form__label form__label--required" htmlFor="confirmEmail">Confirm email</label>
				<input
					className={`form__input${formErrors.confirmEmail.length ? " form__input--error" : ""}`}
					id="confirmEmail"
					placeholder="Confirm email"
					type="text"
					value={formValues.confirmEmail}
					onBlur={() => validateFieldRules(
						"confirmEmail",
						formValues.confirmEmail,
						fieldsValidationRules.confirmEmail,
						setFormErrors,
						formValues
					)}
					onChange={e => setFormValues({field: "confirmEmail", value: e.target.value})}/>
				<em role="alert" className="form__error">{formErrors.confirmEmail.join(" - ")}</em>
			</div>
			<button className="form__button" type="submit">Submit</button>
		</form>
	);
};