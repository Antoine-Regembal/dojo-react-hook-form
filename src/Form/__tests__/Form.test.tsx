import {render} from "@testing-library/react";
import {Form} from "../Form";

describe("Form", function () {
	describe("snapshots", function () {
		it("should match empty form", function () {
			const {container} = render(<Form />);

			expect(container).toMatchSnapshot();
		});
	});
});
