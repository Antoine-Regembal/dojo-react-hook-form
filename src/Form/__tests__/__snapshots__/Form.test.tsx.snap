// Vitest Snapshot v1

exports[`Form > snapshots > should match empty form 1`] = `
<div>
  <form
    class="form form--required"
  >
    <div
      class="input-wrapper"
    >
      <label
        class="form__label form__label--required"
        for="firstname"
      >
        Firstname
      </label>
      <input
        class="form__input"
        id="firstname"
        placeholder="Firstname"
        type="text"
        value=""
      />
      <em
        class="form__error"
        role="alert"
      />
    </div>
    <div
      class="input-wrapper"
    >
      <label
        class="form__label form__label--required"
        for="lastname"
      >
        Lastname
      </label>
      <input
        class="form__input"
        id="lastname"
        placeholder="Lastname"
        type="text"
        value=""
      />
      <em
        class="form__error"
        role="alert"
      />
    </div>
    <div
      class="input-wrapper"
    >
      <label
        class="form__label form__label--required"
        for="email"
      >
        Email
      </label>
      <input
        class="form__input"
        id="email"
        placeholder="Email"
        type="text"
        value=""
      />
      <em
        class="form__error"
        role="alert"
      />
    </div>
    <div
      class="input-wrapper"
    >
      <label
        class="form__label form__label--required"
        for="confirmEmail"
      >
        Confirm email
      </label>
      <input
        class="form__input"
        id="confirmEmail"
        placeholder="Confirm email"
        type="text"
        value=""
      />
      <em
        class="form__error"
        role="alert"
      />
    </div>
    <button
      class="form__button"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
